DOC-2862: Revise query execution modes section for clarity [4.3.0] (#320)
tigergraph/gsql-docs: modules/querying/pages/query-modes.adoc

Query Structure Options:

* Stored Procedure (procedural and saved)
* Anonymous (procedural and ad hoc)
* Non-Procedural (ad hoc)

=== Stored Procedure Queries

> Originally, all GSQL queries were *stored procedures*.

Reviewer: ```suggestion
```

A stored procedure is a named sequence of executable statements that is *saved*, and then executing by calling it by name.

Key characteristics:

* Supports loops and conditional flow to implement algorithms and analytics
* Accepts input parameters
* Has a header and body (bounded by curly braces)

> * Output must be explicitly requested with PRINT statements

Reviewer: ```suggestion
* Output is explicitly specified with PRINT statements
```

> NOTE: Since this was the only query structure format for some time, we will often simply refer to this choice as a *GSQL query*.

Reviewer: ```suggestion
```

[IMPORTANT]
====

> Stored procedural queries that are installed before running generally execute faster and more effiiently.

Reviewer: ```suggestion
Stored procedural queries that are installed before running generally execute faster and more efficiently.
```

|===

== Query Execution Choices

When you run a query in TigerGraph, you make several independent choices.
Each choice affects how the query runs, how it performs, and how it uses system resources.

You can think of query execution as four simple questions:

1. What kind of query are you running?
2. Do you install it or interpret it?
3. If you interpret it, which engine runs it?
4. If you install it, does it run across the cluster?

=== 1. What kind of query are you running?

This choice determines the structure of your query.

* Named query (stored procedure):
You create and save this query using `CREATE QUERY`.

* Anonymous query:

> You run this query once without saving it.

Reviewer: ```suggestion
You run this query at the same time that you define it.
Because it is not saved, it is not given a name (hence, anonymous).
```

* Declarative (non-procedural) query:
You write this simplified query without using `CREATE`, `RUN`, or `INTERPRET`.

=== 2. Do you install it or interpret it?

This choice determines how TigerGraph executes your query.

* INSTALL + RUN:
TigerGraph compiles and optimizes the query before running it.
This provides the best performance.

* INTERPRET:
TigerGraph runs the query immediately without installation.

> This is useful for development and testing.

Reviewer: ```suggestion
This is useful for development and functional testing.
```
Not all testing. If you were doing performance testing, you would want to use the same querying approach that would be applied in real use.

=== Install and run a stored procedure query

> The most performant way to run a GSQL query is to create, install it, and run it.  During installation, TigerGraph performs optimizations that are not available in interpreted mode.

Reviewer: ```suggestion
The most performant way to run a GSQL query is to create, install, and run it.  During installation, TigerGraph performs optimizations that are not available in interpreted mode.
```

[NOTE]
====

> When your application builds queries dynamically, the REST API or pyTigerGraph library is often the best fit.

Reviewer: This note seems to have gotten misplaced. It should be with the section on APIs.

====

==== Interpret a non-procedural query

Introduced in 4.2, this method lets you run simple queries directly in the GSQL shell.

There is no `RUN` or `INTERPRET` command. You simply submit the query.

[console]
----
SELECT m FROM (m:Movie) WHERE m.year == 1939 ORDER BY m.boxoffice DESC LIMIT 10
----

For more examples, see the link:https://github.com/tigergraph/ecosys/blob/master/tutorials/GSQL.md#1-block-query-examples[1-Block Query Examples] in the GSQL V3 tutorial.

== Runtime Engine (Interpreted Queries Only)

If you choose `INTERPRET`, TigerGraph executes the query using an interpreted engine.

> Starting in 4.3.0, interpreted queries can run using one of two engines:

Reviewer: ```suggestion
Starting in 4.3, interpreted queries can run using one of two engines:
```
